Kernel/FileSystem/FATFS/FileSystem.cpp

```diff
@@ -34,46 +105,128 @@ ErrorOr<void> FATFS::initialize_while_locked()
+        if (ebpb_version == DOSBIOSParameterBlockVersion::DOS_BPB_7) {
+            DOS7BIOSParameterBlock const* dos7_boot_record = ebpb.get_dos7_bpb();
+            dbgln("FATFS: EBPB: DOS 7");
+            dbgln("FATFS: flags: {}", dos7_boot_record->flags);
+            dbgln("FATFS: fat_version: {}", dos7_boot_record->fat_version);
+            dbgln("FATFS: root_directory_cluster: {}", dos7_boot_record->root_directory_cluster);
+            dbgln("FATFS: fs_info_sector: {}", dos7_boot_record->fs_info_sector);
+            dbgln("FATFS: backup_boot_sector: {}", dos7_boot_record->backup_boot_sector);
+            dbgln("FATFS: drive_number: {}", dos7_boot_record->drive_number);
+            dbgln("FATFS: volume_id: {}", dos7_boot_record->volume_id);
+        } else if (ebpb_version == DOSBIOSParameterBlockVersion::DOS_BPB_3 || ebpb_version == DOSBIOSParameterBlockVersion::DOS_BPB_4) {
+            DOS4BIOSParameterBlock const* dos4_boot_record = ebpb.get_dos4_bpb();
+            if (ebpb_version == DOSBIOSParameterBlockVersion::DOS_BPB_3) {
+                dbgln("FATFS: EBPB: DOS 3.4");
+            } else if (ebpb_version == DOSBIOSParameterBlockVersion::DOS_BPB_4) {
+                dbgln("FATFS: EBPB: DOS 4");
+            }
+            dbgln("FATFS: drive_number: {}", dos4_boot_record->drive_number);
+            dbgln("FATFS: flags: {}", dos4_boot_record->flags);
+            dbgln("FATFS: volume_id: {}", dos4_boot_record->volume_id);
+
+            // volume_label_string and file_system_type are only valid when
+            // ebpb_version == DOSBIOSParameterBlockVersion::DOS4.
+        }
     }
 
-    if (boot_record()->signature != signature_1 && boot_record()->signature != signature_2) {
+    if (ebpb.signature() != signature_1 && ebpb.signature() != signature_2) {
         dbgln("FATFS: Invalid signature");
         return EINVAL;
     }
 
-    m_device_block_size = boot_record()->bytes_per_sector;
+    // The number of data area sectors is what DOS/Windows used to determine
+    // if a partition was a FAT12, FAT16, or FAT32 file system.
+    // From "FAT Type Determination" section of Microsoft FAT Specification
+    // (fatgen103.doc):
+    //     The FAT type—one of FAT12, FAT16, or FAT32—is determined by the count
+    //     of clusters on the volume and nothing else.
+    //
+    // The following calculations are based on the equations provided in this
+    // section.
+
+    // "RootDirSectors" from MS FAT Specification.
+    u32 root_directory_sectors = (
+                                     // Number of bytes occupied by root directory area (0 on FAT32):
+                                     (ebpb.block->root_directory_entry_count * sizeof(FATEntry)) +
+                                     // Add bytes to fill final sector (ie, round up):
+                                     (ebpb.block->bytes_per_sector - 1))
+        /
+        // Convert byte count to sector count.
+        ebpb.block->bytes_per_sector;
```

Comment:
This is formatted... very strangely. Can we put the comments above the math instead? clang-format did a number on the readability of this block.